# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0 late days

In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import KFold, StratifiedKFold
import warnings

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
slime_df = pd.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
display(slime_df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
display(slime_df.info())

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


None

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = slime_df[["Temperature °C", "Mols KCL"]]
y = slime_df["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
linear = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = X_test.head(1)
print("Predicted: " + str(linear.predict(sample)))
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("Coefficient of determination: " + str(linear.score(X_test, y_test)))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficients: " + str(linear.coef_))
print("Intercept: " + str(linear.intercept_))

Predicted: [235911.1927226]
Coefficient of determination: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340787


Write the linear equation of a slime: $h(x) = -409391.47958 + 866.14641x_1 + 1032.69506x_2$

Report on score and explain meaning: This model has an $R^2$ score (coefficient of determination) of 0.855. The coefficient of determination is the proportion of variability in the labels that is accounted for by the features. A value of 0.855 is high, and means that 85.5% of the variability in the size is accounted for by the features.

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kfold = KFold(n_splits=5, random_state=42, shuffle=True)
linear_cv = cross_val_score(LinearRegression(), X, y, cv=kfold)
print(linear_cv)
print("Average: " + str(linear_cv.mean()) + " Standard deviation: " + str(linear_cv.std()))
# Report on their finding and their significance

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average: 0.8597294202684646 Standard deviation: 0.018387737139306446


Write findings here: The average $R^2$ score is 0.859 across the 5 splits, with a standard deviation of 0.018. The average is very close to the score from part 3, meaning we can interpret it the same way - the score is high, and across the 5 splits an average of 85.9% of the variability in the size was accounted for by the features. The standard deviation is low, meaning that the $R^2$ score was fairly consistent across the 5 splits.

## 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly_X = PolynomialFeatures(degree=2).fit_transform(X)
poly_cv = cross_val_score(LinearRegression(), poly_X, y, cv=kfold)
print(poly_cv)
print("Average: " + str(poly_cv.mean()) + " Standard deviation: " + str(poly_cv.std()))
# Report on the metrics and output the resultant equation as you did in Part 3.
poly_X_train, poly_X_test, poly_y_train, poly_y_test = train_test_split(poly_X, y, test_size=0.1, random_state=42)
poly = LinearRegression().fit(poly_X_train, poly_y_train)
print("Coefficient of determination: " + str(poly.score(poly_X_test, poly_y_test)))
print("Coefficients: " + str(poly.coef_))
print("Intercept: " + str(poly.intercept_))

[1. 1. 1. 1. 1.]
Average: 1.0 Standard deviation: 0.0
Coefficient of determination: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195197e-07  1.26480463e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477513317018747e-05


Write the polynomial equation of a slime: $h(x) = 2.04775*10^{-5} + 1.2*10^{1}x_1 + -1.27195*10^{-7}x_2 + 1.26480*10^{-11}x_1^2 + 2x_1x_2 + 2.85714*10^{-2}x_2^2$

Report on the score and interpret:
This model has an $R^2$ score (coefficient of determination) of 1. The coefficient of determination is the proportion of variability in the labels that is accounted for by the features. A value of 1 is the highest possible, and means that all of the variability in the size is accounted for by the features, meaning that the model is a perfect fit to the data.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [7]:
# Load the dataset. Then train and evaluate the classification models.
ckd_df = pd.read_csv("ckd_feature_subset.csv")

X = ckd_df[["age", "bp", "wbcc", "appet_poor", "appet_good", "rbcc"]]
y = ckd_df["Target_ckd"]

skfold = StratifiedKFold(n_splits=5, random_state=42, shuffle=True)

ckd_df

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
...,...,...,...,...,...,...,...
148,0.181818,0.333333,0.075221,0,1,0.457627,0
149,0.207792,0.166667,0.181416,0,1,0.728814,0
150,0.493506,0.166667,0.172566,0,1,0.711864,0
151,0.675325,0.500000,0.128319,0,1,0.745763,0


In [8]:
#Logistic Regression
logistic_cv = cross_val_score(LogisticRegression(), X, y, cv=skfold)
print(logistic_cv)
print("Logistic Regression Average: " + str(logistic_cv.mean()) + " Standard deviation: " + str(logistic_cv.std()))

[0.90322581 0.87096774 0.87096774 0.8        0.86666667]
Logistic Regression Average: 0.8623655913978496 Standard deviation: 0.0338393776563801


In [9]:
#Support Vector Machines
SVM_cv = cross_val_score(SVC(), X, y, cv=skfold)
print(SVM_cv)
print("Support Vector Machines Average: " + str(SVM_cv.mean()) + " Standard deviation: " + str(SVM_cv.std()))

[0.93548387 1.         0.96774194 0.86666667 0.9       ]
Support Vector Machines Average: 0.9339784946236559 Standard deviation: 0.04730009629719232


In [10]:
#k-Nearest Neighbors
KNN_cv = cross_val_score(KNeighborsClassifier(), X, y, cv=skfold)
print(KNN_cv)
print("k-Nearest Neighbors Average: " + str(KNN_cv.mean()) + " Standard deviation: " + str(KNN_cv.std()))

[0.96774194 1.         0.93548387 0.86666667 0.9       ]
k-Nearest Neighbors Average: 0.9339784946236559 Standard deviation: 0.04730009629719232


In [11]:
#Neural Network
base_NN_cv = cross_val_score(MLPClassifier(random_state=42), X, y, cv=skfold)
print(base_NN_cv)
print("Neural Network Average: " + str(base_NN_cv.mean()) + " Standard deviation: " + str(base_NN_cv.std()))

c:\Users\erisc\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\erisc\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\erisc\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\erisc\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[0.90322581 0.90322581 0.90322581 0.8        0.9       ]
Neural Network Average: 0.8819354838709679 Standard deviation: 0.040986787546499565


c:\Users\erisc\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

| Model | Average | Standard Deviation |
|---|---|---|
| Logistic Regression | 0.8623655913978496 | 0.0338393776563801 |
| Support Vector Machines | 0.9339784946236559 | 0.04730009629719232 |
| k-Nearest Neighbors | 0.9339784946236559 | 0.04730009629719232 |
| Neural Network | 0.8819354838709679 | 0.040986787546499565 |

All models used the default settings. Support vector machines and k-nearest neighbors tied for the highest average accuracy at 0.93, and also had the same standard deviation of 0.047. None of the models performed very poorly, with the worst being logistic regression with an average accuracy of 0.86. Support vector machines and k-nearest neighbors are both better at handling non-linear relationships than logistic regression, likely explaining why those two performed better. Neural networks perform better with large datasets (especially with the solver='adam' default parameter, as explained in the sklearn documentation for MLPClassifier), and this dataset is small, likely explaining why the support vector machines and k-nearest neighbors models did better in this case.

In [12]:
# Experiments with Neural Network.
warnings.filterwarnings("ignore")

for solver in ['adam', "lbfgs", "sgd"]:
    for activation in ['identity', 'logistic', 'tanh', 'relu']:
        print("\nSettings: solver=" + solver + ", activation=" + activation)
        new_NN_cv = cross_val_score(MLPClassifier(solver=solver, activation=activation, random_state=42), X, y, cv=skfold)
        print("Average: " + str(new_NN_cv.mean()) + " Standard deviation: " + str(new_NN_cv.std()))


Settings: solver=adam, activation=identity
Average: 0.8752688172043012 Standard deviation: 0.04020987833985457

Settings: solver=adam, activation=logistic
Average: 0.7320430107526882 Standard deviation: 0.0235815056548349

Settings: solver=adam, activation=tanh
Average: 0.8688172043010752 Standard deviation: 0.037717266698311804

Settings: solver=adam, activation=relu
Average: 0.8819354838709679 Standard deviation: 0.040986787546499565

Settings: solver=lbfgs, activation=identity
Average: 0.9410752688172044 Standard deviation: 0.04783387446954734

Settings: solver=lbfgs, activation=logistic
Average: 0.9866666666666667 Standard deviation: 0.016329931618554516

Settings: solver=lbfgs, activation=tanh
Average: 0.9737634408602152 Standard deviation: 0.024260123454146035

Settings: solver=lbfgs, activation=relu
Average: 0.9737634408602152 Standard deviation: 0.024260123454146035

Settings: solver=sgd, activation=identity
Average: 0.843010752688172 Standard deviation: 0.025409127720846317



## Results and Conclusion for Neural Network Experiments

| Solver | Activation | Average          | Standard Deviation   |
|--------|------------|-------------------|-----------------------|
| adam   | identity   | 0.8752688172043012| 0.04020987833985457   |
| adam   | logistic   | 0.7320430107526882| 0.0235815056548349    |
| adam   | tanh       | 0.8688172043010752| 0.037717266698311804  |
| adam   | relu       | 0.8819354838709679| 0.040986787546499565  |
| lbfgs  | identity   | 0.9410752688172044| 0.04783387446954734   |
| lbfgs  | logistic   | 0.9866666666666667| 0.016329931618554516  |
| lbfgs  | tanh       | 0.9737634408602152| 0.024260123454146035  |
| lbfgs  | relu       | 0.9737634408602152| 0.024260123454146035  |
| sgd    | identity   | 0.843010752688172 | 0.025409127720846317  |
| sgd    | logistic   | 0.7255913978494624| 0.013367974424017224  |
| sgd    | tanh       | 0.8105376344086022| 0.03721362304743789   |
| sgd    | relu       | 0.7387096774193548| 0.027405647005718688  |

The parameter than made the biggest difference in performance was the solver. The lbfgs solver performed much better than the other two, with an average accuracy above 0.9 for all activation settings. This makes sense, since the lbfgs solver is recommended for small datasets. Interestingly, the activation parameter had different effects with different solvers. For the adam solver, the relu activation function performed best with a 0.88 average, with identity and tanh closely following, and logistic performing worse at a 0.73 average. For the lbfgs solver, logistic performed best with a 0.99 average, with tanh and relu closely following, and identity further behind at a 0.94 average. For the sgd solver, identity performed best with a 0.84 average, with tanh slightly behind, and relu and logistic performing worse, with logistic being the worst at a 0.73 average.